sw/device/lib/dif/dif_hmac.c

```diff
@@ -0,0 +1,400 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/dif/dif_hmac.h"
+
+#include "sw/device/lib/base/bitfield.h"
+#include "sw/device/lib/base/memory.h"
+#include "sw/device/lib/base/mmio.h"
+
+#include "hmac_regs.h"  // Generated.
+
+/**
+ * Load `key` into `hmac`.
+ */
+static void load_key(const dif_hmac_t *hmac, const uint8_t *key) {
+  mmio_region_memcpy_to_mmio32(hmac->base_addr, HMAC_KEY0_REG_OFFSET, key,
+                               HMAC_PARAM_NUMWORDS * sizeof(uint32_t));
+}
+
+/**
+ * Read the status register from `hmac`.
+ *
+ * @param hmac The HMAC device to read the status register from.
+ * @return The contents of hmac.STATUS.
+ */
+static uint32_t get_status(const dif_hmac_t *hmac) {
+  return mmio_region_read32(hmac->base_addr, HMAC_STATUS_REG_OFFSET);
+}
+
+/**
+ * Returns the number of entries in the FIFO of `hmac`. If the FIFO is empty,
+ * this function will return 0, and if the FIFO is full, this funciton will
+ * return `HMAC_FIFO_MAX`.
+ *
+ * @param hmac The HMAC device to check the FIFO size of.
+ * @return The number of entries in the HMAC FIFO.
+ */
+static uint32_t get_fifo_entry_count(const dif_hmac_t *hmac) {
+  return bitfield_get_field32(get_status(hmac),
+                              (bitfield_field32_t){
+                                  .mask = HMAC_STATUS_FIFO_DEPTH_MASK,
+                                  .index = HMAC_STATUS_FIFO_DEPTH_OFFSET,
+                              });
+}
+
+/**
+ * A helper funciton for calculating `HMAC_FIFO_MAX` - `get_fifo_entry_count()`.
+ */
+static uint32_t get_fifo_available_space(const dif_hmac_t *hmac) {
+  return HMAC_MSG_FIFO_SIZE_WORDS - get_fifo_entry_count(hmac);
+}
+
+/**
+ * Convert from a `dif_hmac_interrupt_t` to the appropriate bit index.
+ * INTR_STATE, INTR_ENABLE, and INTR_TEST registers have the same bit offset.
+ */
+static bool irq_bit_offset_get(dif_hmac_interrupt_t irq_type,
```

Comment:
This function (and other error-like returning functions) probably wants to be marked as DIF_WARN_UNUSED_RESULT. You should do this in a followup. (Also, might want to make this return `kBadArg` instead of `false`...?)